workshop/exercise-6/README.md

```diff
@@ -1 +1,117 @@
+
+### Canary Deployment
+In `Canary Deployments`, newer versions of services are incrementally rolled out to minizmize the risk and impact of any bugs introduced by the newer version.  To prevent Istio from performing the default routing behavior between the original and modernized guestbook service version, the following rules should be defined:
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: virtual-service-guestbook
+spec:
+  hosts:
+  - guestbook
+  http:
+  - route:
+    - destination:
+        host: guestbook
+        subset: v1
+```
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: DestinationRule
+metadata:
+  name: destination-rule-guestbook
+spec:
+  host: guestbook
+  subsets:
+  - name: v1
+    labels:
+      version: v1
+  - name: v2
+    labels:
+      version: v2
+```
+
+The `VirtualService` defines a rule that captures all HTTP traffic going to hostname `guestbook` and routes that traffic pods of the guestbook service with label "version: v1".  A subset or version of a route destination is identified with a reference to a named service subset which must be declared in a corresponding `DestinationRule`.
+
+To begin incrementally routing traffic to the newer version of the guestbook service, the original `VirtualService` rule can be modified to following:
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: virtual-service-guestbook
+spec:
+  hosts:
+  - guestbook
+  http:
+    - route:
+      - destination:
+          host: guestbook
+          subset: v1
+        weight: 80
+      - destination:
+          host: guestbook
+          subset: v2
+        weight: 20
+```
+
+In the modified rule, the routed traffic is split between two different subsets of the guestbook service.  In this manner, traffic to the modernized version 2 of guestbook is controlled on a percentage basis to limit the impact of any undiscovered bugs.  This rule can be modified over time until eventually all traffic is directed to the newer version of the service.
+
+### A/B Testing with Istio
+A/B testing is a method of performing identical tests against two separate service versions in order to determine which performs better.  The original `VirtualService` rule above can be modified to the following:
```

Comment:
would you recommend A/B testing first before canary?

Reply:
I had just followed the items as they appeared in the IBM box note for the course.  Yes, makes sense to adjust the ordering.